ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp

```diff
@@ -1,300 +1,301 @@
+        Datadog::UploaderBuilder::set_tag(key, val);
 }
 void
-ddup_config_sample_type(unsigned int type)
+ddup_config_sample_type(unsigned int _type)
 {
-    profile_builder.add_type(type);
+    Datadog::SampleManager::add_type(_type);
 }
 void
 ddup_config_max_nframes(int max_nframes)
 {
-    if (max_nframes > 0)
-        profile_builder.set_max_nframes(max_nframes);
-}
-
-#if DDUP_BACKTRACE_ENABLE
-#include <csignal>
-#include <cxxabi.h>
-#include <execinfo.h>
-
-inline static void
-print_backtrace()
-{
-    constexpr int max_frames = 128;
-    void* frames[max_frames];
-    int num_frames = backtrace(frames, max_frames);
-    char** symbols = backtrace_symbols(frames, num_frames);
-
-    std::cerr << "Backtrace:\n";
-    for (int i = 0; i < num_frames; ++i) {
-        std::string symbol(symbols[i]);
-        std::size_t start = symbol.find_first_of('_');
-        std::size_t end = symbol.find_first_of(' ', start);
-
-        if (start != std::string::npos && end != std::string::npos) {
-            std::string mangled_name = symbol.substr(start, end - start);
-            int status = -1;
-            char* demangled_name = abi::__cxa_demangle(mangled_name.c_str(), nullptr, nullptr, &status);
-            if (status == 0) {
-                symbol.replace(start, end - start, demangled_name);
-                free(demangled_name);
-            }
-        }
-        std::cerr << symbol << std::endl;
-    }
-    std::cerr << std::endl;
-
-    free(symbols);
+    Datadog::SampleManager::set_max_nframes(max_nframes);
 }
 
-static void
-sigsegv_handler(int sig, siginfo_t* si, void* uc)
+bool
+ddup_is_initialized()
 {
-    (void)uc;
-    print_backtrace();
-    exit(-1);
+    return is_ddup_initialized;
 }
 
-#endif
 void
 ddup_init()
 {
-    if (!is_initialized) {
-#if DDUP_BACKTRACE_ENABLE
-        // Install segfault handler
-        struct sigaction sigaction_handlers = {};
-        sigaction_handlers.sa_sigaction = sigsegv_handler;
-        sigaction_handlers.sa_flags = SA_SIGINFO;
-        sigaction(SIGSEGV, &(sigaction_handlers), NULL);
-#endif
-
-        g_profile_real[0] = profile_builder.build_ptr();
-        g_profile_real[1] = profile_builder.build_ptr();
-        g_profile = g_profile_real[g_prof_flag];
-        g_uploader = uploader_builder.build_ptr();
-        is_initialized = true;
+    static std::atomic<int> initialized_count{ 0 };
+    static bool initialized = []() {
+        // install the ddup_fork_handler for pthread_atfork
+        // Right now, only do things in the child _after_ fork
+        pthread_atfork(ddup_prefork, ddup_postfork_parent, ddup_postfork_child);
+
+        // Set the global initialization flag
+        is_ddup_initialized = true;
+        return true;
+    }();
+
+    initialized_count.fetch_add(static_cast<int>(initialized));
+    if (initialized_count > 1) {
+        std::cerr << "ddup_init() called " << initialized_count << " times" << std::endl;
     }
 }
 
-void
-ddup_start_sample()
+unsigned int
```

Comment:
Why not just return the actual enum?  Operating over the enum would prevent most of these casts

Reply:
Factored, but discussed the challenges of presenting an enum class across an FFI boundary, especially when that enum is used to index as is the case here.  AFAIK, there's no way to avoid the check at the boundaries.